packages/provider/src/MultiChainProvider.test.ts

```diff
@@ -0,0 +1,404 @@
+      }));
+
+      await provider.request({
+        chainId: 'eip155:1',
+        request: {
+          method: 'eth_accounts',
+        },
+      });
+
+      expect(request).toHaveBeenNthCalledWith(2, {
+        method: 'wallet_multiChainRequestHack',
+        params: {
+          id: expect.any(String),
+          jsonrpc: '2.0',
+          method: 'caip_request',
+          params: {
+            chainId: 'eip155:1',
+            request: {
+              method: 'eth_accounts',
+              params: undefined,
+            },
+          },
+        },
+      });
+    });
+
+    it('throws on JSON-RPC errors', async () => {
+      const provider = await getProvider();
+
+      const request = ethereum.request as jest.MockedFunction<
+        typeof ethereum.request
+      >;
+
+      request.mockImplementation(async () => ({
+        jsonrpc: '2.0',
+        id: 1,
+        error: {
+          code: -1,
+          message: 'foo',
+        },
+      }));
+
+      await expect(
+        provider.request({
+          chainId: 'eip155:1',
+          request: {
+            method: 'eth_accounts',
+          },
+        }),
+      ).rejects.toThrow(
+        'Not a successful JSON-RPC response: At path: result -- Expected the value to satisfy a union of `literal | boolean | number | string | lazy | lazy`, but received: undefined.',
+      );
+    });
+  });
+
+  describe('events', () => {
+    it('emits a "session_delete" when the wallet disconnects', async () => {
+      const provider = await getProvider();
+      expect(provider.isConnected).toBe(true);
+
+      const listener = jest.fn();
+      provider.on('session_delete', listener);
+
+      const on = ethereum.on as jest.MockedFunction<typeof ethereum.on>;
+      const [event, callback] = on.mock.calls[0];
+
+      expect(event).toBe('multichainHack_metamask_disconnect');
+      callback();
+
+      expect(provider.isConnected).toBe(false);
+      expect(listener).toHaveBeenCalledTimes(1);
+    });
+
+    it('emits a "session_event" when the wallet sends an event', async () => {
+      const provider = await getProvider();
+      expect(provider.isConnected).toBe(true);
+
+      const listener = jest.fn();
+      provider.on('session_event', listener);
+
+      const on = ethereum.on as jest.MockedFunction<typeof ethereum.on>;
+      const [event, callback] = on.mock.calls[1];
+
+      expect(event).toBe('multichainHack_metamask_event');
+      callback({
+        method: 'multichainHack_metamask_event',
+        params: {
+          chainId: 'eip155:1',
+          event: {
+            name: 'foo',
+            data: 'bar',
+          },
+        },
+      });
+
+      expect(listener).toHaveBeenCalledTimes(1);
+      expect(listener).toHaveBeenCalledWith({
+        params: {
+          chainId: 'eip155:1',
+          event: {
+            name: 'foo',
+            data: 'bar',
+          },
+        },
+      });
+    });
+
+    it('verifies the event arguments', async () => {
```

Comment:
Is that a user facing requirement?

Reply:
Nope, but regardless a good thing to test I think.